pkg/user/manager/owncloudsql/accounts/accounts.go

```diff
@@ -0,0 +1,225 @@
+		appctx.GetLogger(ctx).Error().Err(err).Msg("could not scan row, skipping")
+		return nil, err
+	}
+
+	return &a, nil
+}
+
+// Scannable describes the interface providing a Scan method
+type Scannable interface {
+	Scan(...interface{}) error
+}
+
+// GetAccountByClaim fetches an account by mail, username or userid
+func (as *Accounts) GetAccountByClaim(ctx context.Context, claim, value string) (*Account, error) {
+	// TODO align supported claims with rest driver and the others, maybe refactor into common mapping
+	var row *sql.Row
+	var where string
+	switch claim {
+	case "mail":
+		where = "WHERE a.email=?"
+	// case "uid":
+	//	claim = m.c.Schema.UIDNumber
+	// case "gid":
+	//	claim = m.c.Schema.GIDNumber
+	case "username":
+		if as.joinUsername {
+			where = "WHERE p.configvalue=?"
+		} else {
+			// use user_id as username
+			where = "WHERE a.user_id=?"
+		}
+	case "userid":
+		if as.joinUUID {
+			where = "WHERE p2.configvalue=?"
+		} else {
+			// use user_id as uuid
+			where = "WHERE a.user_id=?"
+		}
+	default:
+		return nil, errors.New("owncloudsql: invalid field " + claim)
+	}
+
+	row = as.db.QueryRowContext(ctx, as.selectSQL+where, value)
+
+	return as.rowToAccount(ctx, row)
+}
+
+func sanitizeWildcards(q string) string {
+	return strings.ReplaceAll(strings.ReplaceAll(q, "%", `\%`), "_", `\_`)
+}
+
+// FindAccounts searches userid, displayname and email using the given query. The Wildcard caracters % and _ are escaped.
+func (as *Accounts) FindAccounts(ctx context.Context, query string) ([]Account, error) {
+	if as.enableMedialSearch {
+		query = "%" + sanitizeWildcards(query) + "%"
+	}
+	// TODO join oc_account_terms
+	where := "WHERE a.user_id LIKE ? OR a.display_name LIKE ? OR a.email LIKE ?"
+	args := []interface{}{query, query, query}
+
+	if as.joinUsername {
+		where += " OR p.configvalue LIKE ?"
+		args = append(args, query)
+	}
+	if as.joinUUID {
+		where += " OR p2.configvalue LIKE ?"
+		args = append(args, query)
+	}
+
+	rows, err := as.db.QueryContext(ctx, as.selectSQL+where, args...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	accounts := []Account{}
+	for rows.Next() {
+		a := Account{}
+		if err := rows.Scan(&a.ID, &a.Email, &a.UserID, &a.DisplayName, &a.Quota, &a.LastLogin, &a.Backend, &a.Home, &a.State, &a.Username, &a.OwnCloudUUID); err != nil {
+			appctx.GetLogger(ctx).Error().Err(err).Msg("could not scan row, skipping")
+			continue
+		}
+		accounts = append(accounts, a)
+	}
+	if err = rows.Err(); err != nil {
+		return nil, err
+	}
+
+	return accounts, nil
+}
+
+// GetAccountGroups lasts the groups for an account
+func (as *Accounts) GetAccountGroups(ctx context.Context, uid string) ([]string, error) {
+	rows, err := as.db.QueryContext(ctx, "SELECT gid FROM oc_group_user WHERE uid=?", uid)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var group string
```

Comment:
This can be moved into the loop to reduce the scope.

Reply:
wouldn't that add an allocation for every iteration? I wanted to reuse the variable, because it will be copied when appending it to groups.

Reply:
No, IIRC there will be only one allocation and the compiler can apply more optimizations since it knows that the value will not be referenced outside of the loop.

Reply:
Here this is not exactly the same code but should behave the same:
```golang
package loop

import (
	"strconv"
)

func declareOutside() []string {
	results := make([]string, 0, 32)
	var s string
	for i := 0; i < 32; i++ {
		s = strconv.Itoa(i)
		results = append(results, s)
	}
	return results
}

func declareInside() []string {
	results := make([]string, 0, 32)
	for i := 0; i < 32; i++ {
		var s string
		s = strconv.Itoa(i)
		results = append(results, s)
	}
	return results
}
```
```
go test -bench=. -benchmem
goos: linux
goarch: amd64
pkg: example.com/loop
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkOutside-8   	 3056689	       395.5 ns/op	     512 B/op	       1 allocs/op
BenchmarkInside-8    	 3067557	       393.0 ns/op	     512 B/op	       1 allocs/op
PASS
ok  	example.com/loop	3.214s
```

Reply:
Oh cool!